### PR TITLE
ESQL: Delay construction of warnings

### PR DESCRIPTION
**Pull request by nik9000**
Delay construction of `Warnings` until they are needed to save memory when evaluating many many many expressions. Most expressions won't use warnings at all and there isn't any need to make registering warnings super duper fast. So let's make the construction lazy to save a little memory. It's like 200 bytes per expression which isn't much, but it's possible to have thousands of expressions in a single query. Abusive, but possible.

This also consolidates all `Warnings` usages to a single `Warnings` class. We had two. We don't need two.

### CODE REVIEW

**Review by mark-vieira — APPROVED** [review]

File: test/external-modules/esql-heap-attack/src/javaRestTest/java/org/elasticsearch/xpack/esql/heap_attack/HeapAttackIT.java [hunk 1]
```diff
@@ -409,9 +409,6 @@ protected void doRun() throws Exception {
                     TimeValue elapsed = TimeValue.timeValueNanos(System.nanoTime() - startedTimeInNanos);
                     logger.info("--> test {} triggering OOM after {}", getTestName(), elapsed);
                     Request triggerOOM = new Request("POST", "/_trigger_out_of_memory");
-                    RequestConfig requestConfig = RequestConfig.custom()
-                        .setSocketTimeout(Math.toIntExact(TimeValue.timeValueMinutes(2).millis()))
-                        .build();
                     client().performRequest(triggerOOM);
```
Comment (ivancea) [inline test/external-modules/esql-heap-attack/src/javaRestTest/java/org/elasticsearch/xpack/esql/heap_attack/HeapAttackIT.java:412]: Is this change related?

**Review by ivancea — APPROVED** [review]
Thanks Nik! The compute Warnings were clearly left half done 😅

File: test/external-modules/esql-heap-attack/src/javaRestTest/java/org/elasticsearch/xpack/esql/heap_attack/HeapAttackIT.java (hunk 1 above)
Comment (nik9000) [reply test/external-modules/esql-heap-attack/src/javaRestTest/java/org/elasticsearch/xpack/esql/heap_attack/HeapAttackIT.java:412]: Sorry. Should have comment on this one. We were totally not using the request config. It was an unused variable and confused me.

**Comment by nik9000** [comment]
@elasticmachine update branch
